File: service/src/main/kotlin/com/hosopy/kotlingrpcsample/grpc/service/GreeterService.kt
package com.hosopy.kotlingrpcsample.grpc.service

import com.hosopy.kotlingrpcsample.greeter.proto.GreeterGrpc
import com.hosopy.kotlingrpcsample.greeter.proto.GreeterOuterClass
import com.hosopy.kotlingrpcsample.grpc.client.TextAnalyzerClient
import io.grpc.stub.StreamObserver
import org.lognet.springboot.grpc.GRpcService

@GRpcService
class GreeterService(
    private val textAnalyzerClient: TextAnalyzerClient
): GreeterGrpc.GreeterImplBase() {
    override fun sayHello(
        request: GreeterOuterClass.HelloRequest,
        responseObserver: StreamObserver<GreeterOuterClass.HelloReply>
    ) {
        val replyBuilder = GreeterOuterClass.HelloReply.newBuilder()
        replyBuilder.message = "Hello " + request.name
        replyBuilder.nameLength = textAnalyzerClient.analyzeText(request.name).length
        responseObserver.onNext(replyBuilder.build())
        responseObserver.onCompleted()
    }
}

File: web/service/GreeterService.ts
import { GreeterPromiseClient } from './grpc/Greeter_grpc_web_pb'
import { HelloRequest } from './grpc/Greeter_pb'

export default class GreeterService {
  constructor(private readonly hostname: string) {
    this.client = new GreeterPromiseClient(hostname, null, null)
  }

  private readonly client: GreeterPromiseClient

  public async sayHello(name: string): Promise<{message: string; nameLength: number}> {
    const request = new HelloRequest()
    request.setName(name)
    const response = await this.client.sayHello(request)
    return Promise.resolve({
      message: response.getMessage(),
      nameLength: response.getNamelength()
    })
  }
}
